File: src/main/java/com/github/sorusclient/client/hud/impl/armor/v1_8_9/ArmorRenderer.java
package com.github.sorusclient.client.hud.impl.armor.v1_8_9;

import com.github.glassmc.loader.GlassLoader;
import com.github.glassmc.loader.Listener;
import com.github.sorusclient.client.adapter.IItem;
import com.github.sorusclient.client.adapter.v1_8_9.Util;
import com.github.sorusclient.client.hud.impl.armor.IArmorRenderer;
import com.github.sorusclient.client.hud.impl.hunger.IHungerRenderer;
import org.lwjgl.opengl.GL11;
import v1_8_9.net.minecraft.client.MinecraftClient;
import v1_8_9.net.minecraft.client.gui.DrawableHelper;
import v1_8_9.net.minecraft.item.Item;
import v1_8_9.net.minecraft.item.ItemStack;
import v1_8_9.net.minecraft.util.Identifier;

public class ArmorRenderer implements Listener, IArmorRenderer {

    @Override
    public void run() {
        GlassLoader.getInstance().registerInterface(IArmorRenderer.class, this);
    }

    @Override
    public void render(IItem item, double x, double y, double scale) {
        ItemStack itemStack = (ItemStack) item.getInner();

        if (itemStack == null) {
            int id = Util.getIdByItemType(item.getType());
            if (id != -1) {
                itemStack = new ItemStack(Item.byRawId(id));
            }
        }

        if (itemStack == null) {
            System.out.println("unknown hotbar item");
            return;
        }

        GL11.glPushMatrix();
        GL11.glTranslated(x, y, 0);
        GL11.glScaled(scale * 18 / 16, scale * 18 / 16, 0);
        MinecraftClient.getInstance().getItemRenderer().renderItem(itemStack, 0 ,0);
        GL11.glPopMatrix();
    }

    @Override
    public void renderArmorPlateBackground(double x, double y, double scale) {
        GL11.glPushMatrix();
        GL11.glTranslated(x, y, 0);
        GL11.glScaled(scale, scale, 1);

        DrawableHelper drawableHelper = new DrawableHelper();
        GL11.glColor4d(1.0, 1.0, 1.0, 1.0);

        MinecraftClient.getInstance().getTextureManager().bindTexture(new Identifier("textures/gui/icons.png"));

        drawableHelper.drawTexture(0, 0, 16, 9, 9, 9);

        GL11.glPopMatrix();
    }

    @Override
    public void renderArmorPlate(double x, double y, double scale, ArmorRenderType armorRenderType) {
        GL11.glPushMatrix();
        GL11.glTranslated(x, y, 0);
        GL11.glScaled(scale, scale, 1);

        DrawableHelper drawableHelper = new DrawableHelper();
        GL11.glColor4d(1.0, 1.0, 1.0, 1.0);

        MinecraftClient.getInstance().getTextureManager().bindTexture(new Identifier("textures/gui/icons.png"));

        if (armorRenderType == ArmorRenderType.FULL) {
            drawableHelper.drawTexture(0, 0, 43, 9, 9, 9);
        } else if (armorRenderType == ArmorRenderType.HALF) {
            drawableHelper.drawTexture(0, 0, 25, 9, 9, 9);
        }

        GL11.glPopMatrix();
    }

}

File: src/main/kotlin/com/github/sorusclient/client/hud/impl/armor/v1_8_9/ArmorRenderer.kt
package com.github.sorusclient.client.hud.impl.armor.v1_8_9

import com.github.glassmc.loader.api.GlassLoader
import com.github.glassmc.loader.api.Listener
import com.github.sorusclient.client.adapter.IItem
import com.github.sorusclient.client.adapter.v1_8_9.Util
import com.github.sorusclient.client.hud.impl.armor.IArmorRenderer
import com.github.sorusclient.client.hud.impl.armor.IArmorRenderer.ArmorRenderType
import org.lwjgl.opengl.GL11
import v1_8_9.net.minecraft.client.MinecraftClient
import v1_8_9.net.minecraft.client.render.Tessellator
import v1_8_9.net.minecraft.client.render.VertexFormats
import v1_8_9.net.minecraft.item.Item
import v1_8_9.net.minecraft.item.ItemStack
import v1_8_9.net.minecraft.util.Identifier

class ArmorRenderer : Listener, IArmorRenderer {
    override fun run() {
        GlassLoader.getInstance().registerInterface(IArmorRenderer::class.java, this)
    }

    override fun render(item: IItem, x: Double, y: Double, scale: Double) {
        var itemStack = item.inner as ItemStack
        if (itemStack == null) {
            val id = Util.getIdByItemType(item.type)
            if (id != -1) {
                itemStack = ItemStack(Item.byRawId(id))
            }
        }
        GL11.glPushMatrix()
        GL11.glTranslated(x, y, 0.0)
        GL11.glScaled(scale * 18 / 16, scale * 18 / 16, 0.0)
        MinecraftClient.getInstance().itemRenderer.renderItem(itemStack, 0, 0)
        GL11.glPopMatrix()
    }

    override fun renderArmorPlateBackground(x: Double, y: Double, scale: Double) {
        GL11.glPushMatrix()
        GL11.glTranslated(x, y, 0.0)
        GL11.glScaled(scale, scale, 1.0)
        GL11.glColor4d(1.0, 1.0, 1.0, 1.0)
        MinecraftClient.getInstance().textureManager.bindTexture(Identifier("textures/gui/icons.png"))
        drawTexture(0, 0, 16.0, 9.0, 9, 9)
        GL11.glPopMatrix()
    }

    override fun renderArmorPlate(x: Double, y: Double, scale: Double, armorRenderType: ArmorRenderType?) {
        GL11.glPushMatrix()
        GL11.glTranslated(x, y, 0.0)
        GL11.glScaled(scale, scale, 1.0)
        GL11.glColor4d(1.0, 1.0, 1.0, 1.0)
        MinecraftClient.getInstance().textureManager.bindTexture(Identifier("textures/gui/icons.png"))
        if (armorRenderType === ArmorRenderType.FULL) {
            drawTexture(0, 0, 43.0, 9.0, 9, 9)
        } else if (armorRenderType === ArmorRenderType.HALF) {
            drawTexture(0, 0, 25.0, 9.0, 9, 9)
        }
        GL11.glPopMatrix()
    }

    private fun drawTexture(var1: Int, var2: Int, var3: Double, var4: Double, var5: Int, var6: Int) {
        val var3 = var3 + 0.1
        val var9 = Tessellator.getInstance()
        val var10 = var9.buffer
        var10.begin(7, VertexFormats.POSITION_TEXTURE)
        var10.vertex((var1 + 0).toDouble(), (var2 + var6).toDouble(), 0.0).texture(((var3 + 0).toFloat() / 256).toDouble(), ((var4 + var6).toFloat() / 256).toDouble()).next()
        var10.vertex((var1 + var5).toDouble(), (var2 + var6).toDouble(), 0.0).texture(((var3 + var5).toFloat() / 256).toDouble(), ((var4 + var6).toFloat() / 256).toDouble()).next()
        var10.vertex((var1 + var5).toDouble(), (var2 + 0).toDouble(), 0.0).texture(((var3 + var5).toFloat() / 256).toDouble(), ((var4 + 0).toFloat() / 256).toDouble()).next()
        var10.vertex((var1 + 0).toDouble(), (var2 + 0).toDouble(), 0.0).texture(((var3 + 0).toFloat() / 256).toDouble(), ((var4 + 0).toFloat() / 256).toDouble()).next()
        var9.draw()
    }

}